scripts/publisher/__main__.py

```diff
@@ -0,0 +1,120 @@
+import argparse
+import subprocess
+import sys
+import os
+from pathlib import Path
+import re
+import tempfile
+
+from packaging import version as packaging_version
+
+from publisher.parser.index_html_parser import IndexHTMLParser
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    required = (("repo-url", "The repository URL where the Python package lives. "
+                             "It start with \"https://\"."),
+                ("github-token", "GitHub token to use for pushing to the index repository."),
+                ("index-name", "Index repository name on GitHub, e.g. "
+                               "\"rporres/python-package-server\"."))
+    optional = (("package-path", "Path to the Python package root.", "."),
+                ("repo-tag", "The tag to publish.", ""),
+                ("target-branch", "The Git branch to which to publish the package.", "master"))
+    required = tuple((p + (None,)) for p in required)
+    for arg, help, default in required + optional:
+        default = os.getenv(arg.replace("-", "_").upper(), default)
+        parser.add_argument("--" + arg, help=help + " Default: \"%s\"." % default, default=default)
+    args = parser.parse_args()
+    missing = False
+    required = set(p[0] for p in required)
+    for k, v in vars(args).items():
+        # required=True will not work here because we need to check the environment variables
+        if k.replace("_", "-") not in required:
+            continue
+        if not v:
+            print("Missing argument --%s / environment variable %s" %
+                  (k.replace("_", "-"), k.upper()), file=sys.stderr)
+            missing = True
+    return args, missing
+
+
+def shell(*args):
+    print(" ".join(args))
+    subprocess.run(args, check=True)
+
+
+def main():
+    args, missing = parse_args()
+    if missing:
+        return 1
+
+    cwd = os.getcwd()
+    os.chdir(args.package_path)
+    cmd = [sys.executable, str(Path(args.package_path) / "setup.py"), "--name", "--version",
+           "--classifiers"]
+    print(" ".join(cmd))
+    metadata = subprocess.check_output(cmd).decode().split("\n")
+    package_name, package_version = metadata[:2]
+    # Normalize name
+    # See https://www.python.org/dev/peps/pep-0503/#normalized-names
+    normalized_package_name = re.sub(r"[-_.]+", "-", package_name).lower()
+    index_file = Path("docs") / normalized_package_name / "index.html"
```

Comment:
The `docs` directory is a convention used by GitHub pages. I found that it was a suitable structure for us but it could be different for others, now that the audience of the script is wider. I would move it into a optional argument, with the empty string as default.

Reply:
Done. I left "docs" by default with the comment that we assume GitHub Pages by default, which is very reasonable for our users.

I had to add `--signature / SIGNATURE` arg because indeed not everybody is source{d} :smile: